core/src/token_info/cached.rs

```diff
@@ -0,0 +1,155 @@
+use super::{TokenBaseInfo, TokenId, TokenInfoFetching};
+
+use anyhow::Result;
+use async_std::sync::RwLock;
+use futures::future::{BoxFuture, FutureExt};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+/**
+ * Implementation of TokenInfoFetching that stores previously fetched information in an in-memory cache for fast retrieval.
+ * TokenIds will always be fetched from the inner layer, as new tokens could be added at any time.
+ */
+struct TokenInfoCache {
+    cache: RwLock<HashMap<TokenId, TokenBaseInfo>>,
+    inner: Arc<dyn TokenInfoFetching>,
+}
+
+impl TokenInfoCache {
+    #[allow(dead_code)]
+    pub fn new(inner: Arc<dyn TokenInfoFetching>) -> Self {
+        Self {
+            cache: RwLock::new(HashMap::new()),
+            inner,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn new_with_cache(
```

Comment:
Hehe, I was suggesting more:
```suggestion
    pub fn with_existing_cache(
```

or 
```suggestion
    pub fn with_cache(
```

Reply:
🤦 of course!